src/Protocol/RecipeMapper.cpp

```diff
@@ -0,0 +1,124 @@
+#include "Globals.h"
+#include "RecipeMapper.h"
+#include "../Root.h"
+
+cRecipeMapper::cRecipeMapper(void)
+{
+	const AString Protocols [] = {"1.12.2"};
+	for (const auto & Protocol: Protocols)
+	{
+		loadRecipes(Protocol);
+	}
+}
+
+
+
+
+
+void cRecipeMapper::loadRecipes(const AString & a_ProtocolVersion)
+{
+	cFile f;
+	if (!f.Open("Protocol/" + a_ProtocolVersion + "/base.recipes.txt", cFile::fmRead))
+	{
+		LOGWARNING("Cannot open file \"Protocol/%s/base.recipes.txt\", no recipe book recipes will be available!", a_ProtocolVersion);
+		return;
+	}
+	AString Everything;
+	if (!f.ReadRestOfFile(Everything))
+	{
+		LOGWARNING("Cannot read file \"Protocol/%s/base.recipes.txt\", no recipe book recipes will be available!", a_ProtocolVersion);
+		return;
+	}
+	f.Close();
+
+	// Split it into lines, then process each line as a single recipe:
+	AStringVector Split = StringSplit(Everything, "\n");
+	m_ProtocolVersionMap[a_ProtocolVersion] = {};
+	auto RecipeNameMap = cRoot::Get()->GetCraftingRecipes()->getRecipeNameMap();
+
+	int LineNum = 1;
+	for (AStringVector::const_iterator itr = Split.begin(); itr != Split.end(); ++itr, ++LineNum)
+	{
+		// Remove anything after a '#' sign and trim away the whitespace:
+		AString Recipe = TrimString(itr->substr(0, itr->find('#')));
+		if (Recipe.empty())
+		{
+			// Empty recipe
+			continue;
+		}
+		AddRecipeLine(a_ProtocolVersion, LineNum, Recipe, RecipeNameMap);
+	}
+	LOG("Loaded %s %zu recipe book", a_ProtocolVersion, m_ProtocolVersionMap[a_ProtocolVersion].size());
+}
+
+
+
+
+
+cRecipeMapper::~cRecipeMapper()
+{
+}
+
+
+
+
+
+void cRecipeMapper::AddRecipeLine(const AString & a_ProtocolVersion, int a_LineNum, const AString & a_RecipeLine, const std::map<AString, UInt32> & a_RecipeNameMap)
+{
+	AStringVector Sides = StringSplit(a_RecipeLine, " ");
+	UInt32 Id;
+	if (Sides.size() != 2)
+	{
+		LOGINFO("Recipe incompletely configured %s", a_RecipeLine);
```

Comment:
Should probably stop processing here as ther might not be a `Sides[1]`.
```suggestion
		LOGINFO("Recipe incompletely configured %s", a_RecipeLine);
		return;
```